cpu/mips32r2_common/thread_arch.c

```diff
@@ -0,0 +1,235 @@
+		 * get routed to debug probe or bootloader handler and output
+		 * via a UART
+		 */
+
+		if (syscall_num == __MIPS_UHI_SYSCALL_NUM) {
+			if (ctx->t2[1] == __MIPS_UHI_WRITE &&
+			   (ctx->a[0] == STDOUT_FILENO || ctx->a[0] == STDERR_FILENO)) {
+				uint32_t status = irq_arch_disable();
+				uart_write(DEBUG_VIA_UART, (uint8_t *)ctx->a[1], ctx->a[2]);
+				ctx->v[0] = ctx->a[2];
+				ctx->epc += 4; /* move PC past the syscall */
+				irq_arch_restore(status);
+				return;
+			} else if (ctx->t2[1] == __MIPS_UHI_FSTAT &&
+				(ctx->a[0] == STDOUT_FILENO || ctx->a[0] == STDERR_FILENO)) {
+				/*
+				 * Printf fstat's the stdout/stderr file so
+				 * fill out a minimal struct stat.
+				 */
+				struct stat *sbuf = (struct stat *)ctx->a[1];
+				sbuf->st_mode = S_IRUSR | S_IWUSR | S_IWGRP;
+				sbuf->st_blksize = BUFSIZ;
+				sbuf->st_dev = 0;
+				sbuf->st_ino = 0;
+				sbuf->st_nlink = 0;
+				sbuf->st_uid = 0;
+				sbuf->st_gid = 0;
+				sbuf->st_rdev = 0;
+				sbuf->st_size = 0;
+				sbuf->st_atime = (time_t) 0;
+				sbuf->st_spare1 = 0;
+				sbuf->st_mtime = (time_t) 0;
+				sbuf->st_spare2 = 0;
+				sbuf->st_ctime = (time_t) 0;
+				sbuf->st_spare3 = 0;
+				sbuf->st_blocks = 0;
+				sbuf->st_spare4[0] = 0;
+				sbuf->st_spare4[1] = 0;
+				/* return 0 */
+				ctx->v[0] = 0;
+				ctx->epc += 4; /* move PC past the syscall */
+				return;
+			}
+		} else
+#endif
+		if (syscall_num == 2) {
+			/*
+			 * Syscall 1 is reserved for UHI.
+			 *
+			 * save the stack pointer in the thread info
+			 * note we want the saved value to include the
+			 * saved off context and the 16 bytes padding.
+			 * Note we cannot use the current sp value as
+			 * the prologue of this function has adjusted it
+			 */
+			sched_active_thread->sp = (char *)(ctx->sp
+					- sizeof(struct gpctx) - 16);
+
+			sched_run();
+
+			new_ctx = (struct gpctx *)((int)sched_active_thread->sp + 16);
+
+			return_instruction = *((unsigned int *)(new_ctx->epc));
+
+			if ((return_instruction & 0xfc00003f) == 0xC) {/* syscall */
+				new_ctx->epc += 4; /* move PC past the syscall */
+			}
+
+			/*
+			 * The toolchain Exception restore code just wholesale copies the
+			 * status register from the context back to the register loosing
+			 * any changes that may have occured, 'status' is really global state
+			 * You dont enable interrupts on one thread and not another...
+			 * So we just copy the current status value into the saved value
+			 * so nothing changes on the restore
+			 */
+
+			new_ctx->status = mips32_get_c0(C0_STATUS);
+
+			asm volatile ("lw    $sp, 0(%0)" : : "r"(&sched_active_thread->sp));
+
+			/*
+			 * Jump straight to the exception restore code
+			 * if we return this functions prologue messes up
+			 * the stack  pointer
+			 */
+			__exception_restore();
+
+			UNREACHABLE();
+		}
+		break;
+
+		/* default: */
+	}
+	/* Pass all other exceptions through to the toolchain handler */
+	__exception_handle(ctx, exception);
+}
+
```

Comment:
Are you sure you pulled the latest version, uncrustify changed all the tabs to spaces ?